File: src/notify.py
import json
import logging
import time
import urllib.error
import urllib.parse
import urllib.request
import uuid

log = logging.getLogger(__name__)

MAX_RETRIES = 3
BACKOFF_BASE = 2


def post_message(
    homeserver: str,
    room_id: str,
    access_token: str,
    plain_body: str,
    html_body: str | None = None,
) -> bool:
    txn_id = str(uuid.uuid4())
    encoded_room = urllib.parse.quote(room_id, safe="")
    url = (
        f"{homeserver.rstrip('/')}/_matrix/client/v3/rooms/"
        f"{encoded_room}/send/m.room.message/{txn_id}"
    )

    content: dict = {"msgtype": "m.text", "body": plain_body}
    if html_body:
        content["format"] = "org.matrix.custom.html"
        content["formatted_body"] = html_body

    data = json.dumps(content).encode()
    req = urllib.request.Request(url, data=data, method="PUT")
    req.add_header("Authorization", f"Bearer {access_token}")
    req.add_header("Content-Type", "application/json")

    for attempt in range(1, MAX_RETRIES + 1):
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                resp_data = json.loads(resp.read())
                log.info("Message sent: %s", resp_data.get("event_id", "?"))
                return True
        except (urllib.error.URLError, urllib.error.HTTPError, OSError) as e:
            if attempt < MAX_RETRIES:
                wait = BACKOFF_BASE ** attempt
                log.warning(
                    "Matrix post failed (attempt %d/%d), retrying in %ds: %s",
                    attempt, MAX_RETRIES, wait, e,
                )
                time.sleep(wait)
            else:
                log.error("Matrix post failed after %d attempts: %s", MAX_RETRIES, e)
    return False


def send(config: dict, plain_body: str, html_body: str | None = None) -> bool:
    matrix = config["matrix"]
    return post_message(
        homeserver=matrix["homeserver"],
        room_id=matrix["room_id"],
        access_token=matrix["access_token"],
        plain_body=plain_body,
        html_body=html_body,
    )

File: notify.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"time"

	"github.com/google/uuid"
)

const (
	maxRetries  = 3
	httpTimeout = 30 * time.Second
)

var matrixHTTPClient = &http.Client{Timeout: httpTimeout}

type matrixMessage struct {
	MsgType       string `json:"msgtype"`
	Body          string `json:"body"`
	Format        string `json:"format,omitempty"`
	FormattedBody string `json:"formatted_body,omitempty"`
}

func PostMessage(ctx context.Context, homeserver, roomID, accessToken, plainBody, htmlBody string) error {
	txnID := uuid.New().String()
	encodedRoom := url.PathEscape(roomID)
	endpoint := fmt.Sprintf("%s/_matrix/client/v3/rooms/%s/send/m.room.message/%s",
		homeserver, encodedRoom, txnID)

	msg := matrixMessage{
		MsgType: "m.text",
		Body:    plainBody,
	}
	if htmlBody != "" {
		msg.Format = "org.matrix.custom.html"
		msg.FormattedBody = htmlBody
	}

	data, _ := json.Marshal(msg)

	var lastErr error
	for attempt := 1; attempt <= maxRetries; attempt++ {
		req, err := http.NewRequestWithContext(ctx, http.MethodPut, endpoint, bytes.NewReader(data))
		if err != nil {
			return fmt.Errorf("failed to create request: %w", err)
		}
		req.Header.Set("Authorization", "Bearer "+accessToken)
		req.Header.Set("Content-Type", "application/json")
		req.GetBody = func() (io.ReadCloser, error) {
			return io.NopCloser(bytes.NewReader(data)), nil
		}

		resp, err := matrixHTTPClient.Do(req)
		if err != nil {
			lastErr = err
			if attempt < maxRetries {
				wait := 1 << attempt // 2, 4
				log.Printf("WARNING: Matrix post failed (attempt %d/%d), retrying in %ds: %v",
					attempt, maxRetries, wait, err)
				select {
				case <-time.After(time.Duration(wait) * time.Second):
				case <-ctx.Done():
					return ctx.Err()
				}
				continue
			}
			return fmt.Errorf("Matrix post failed after %d attempts: %w", maxRetries, lastErr)
		}
		io.Copy(io.Discard, resp.Body)
		resp.Body.Close()

		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			log.Printf("INFO: Message sent (status %d)", resp.StatusCode)
			return nil
		}

		lastErr = fmt.Errorf("HTTP %d", resp.StatusCode)
		if attempt < maxRetries {
			wait := 1 << attempt
			log.Printf("WARNING: Matrix post returned %d (attempt %d/%d), retrying in %ds",
				resp.StatusCode, attempt, maxRetries, wait)
			select {
			case <-time.After(time.Duration(wait) * time.Second):
			case <-ctx.Done():
				return ctx.Err()
			}
		}
	}
	return fmt.Errorf("Matrix post failed after %d attempts: %w", maxRetries, lastErr)
}
